zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/api/factory/RealmAwareZkClientFactory.java

```diff
@@ -19,5 +19,14 @@
  * under the License.
  */
 
+import org.apache.helix.zookeeper.api.api.client.RealmAwareZkClient;
+
+
+/**
+ * Creates an instance of RealmAwareZkClient.
+ */
 public interface RealmAwareZkClientFactory {
+
+  RealmAwareZkClient buildZkClient(RealmAwareZkClient.MODE mode,
+      RealmAwareZkClient.RealmAwareZkClientConfig config);
```

Comment:
Just a note, this means that we need to use the RealmAwareZkClientConfig to genereate HelixZkClient.ZkConnectionConfig. Needs validate.

Reply:
Another high level question, in the scope of dedicatedZkClient or sharedZkClient discussion, all the create ZkClient already are RealAware, right? we don't really need to have a mode for user to specify.

The mode can be kept internally for the client to know. In fact, even this (internally keep the type) may not be needed as we have new dedicatedZkClient and new sharedZkClient which wraps sharedZkClient and zkclient as of now. The type information is already encode in the class design. The beauty of OO inheritance.

Reply:
@kaisun2000 
1. Yes, we should add the sessionTimeout field to this. I will add it for completeness (this was a draft, so I just wanted to first give you an idea). :)
2. Yes and no.
- RealmAwareZkClients produced by RealmAwareZkClientFactory are all single-realm mode. So yes, mode might not be required as a method parameter for `buildZkClient()`.
- We should still keep this field internally because not all RealmAwareZkClients are single-realm mode. For example, if you create a FederatedZkClient instance, it should still have a `mode = MULTI_REALM`.
So yes, let's follow your suggestion to remove the mode parameter from here, but we should keep it as an internal variable that we could expose to the user (for instance, RealmAwareZkClient should support an interface method like `getMode()` that returns either `SINGLE_REALM` or `MULTI_REALM`)